native/src/main/java/io/ballerina/stdlib/nats/jetstream/listener/StreamMessageHandler.java

```diff
@@ -0,0 +1,193 @@
+
+    public StreamMessageHandler(BObject service, Runtime runtime, String connectedUrl, boolean autoAck) {
+        this.service = service;
+        this.runtime = runtime;
+        this.autoAck = autoAck;
+        this.connectedUrl = connectedUrl;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void onMessage(Message msg) {
+        BMap<BString, Object> msgRecord = ValueCreator.createRecordValue(
+                Utils.getModule(), Constants.STREAM_MESSAGE);
+        Object[] msgRecordValues = new Object[2];
+
+        msgRecordValues[0] = StringUtils.fromString(msg.getSubject());
+        msgRecordValues[1] = ValueCreator.createArrayValue(msg.getData());
+
+        BMap<BString, Object> populatedMsgRecord = ValueCreator.createRecordValue(msgRecord, msgRecordValues);
+
+        BObject callerObj = ValueCreator.createObjectValue(Utils.getModule(), Constants.STREAM_CALLER);
+        callerObj.addNativeData(Constants.JET_STREAM_MESSAGE, msg);
+        callerObj.addNativeData(Constants.JET_STREAM_AUTO_ACK, autoAck);
+
+        MethodType onMessageResource = getAttachedFunctionType(service, "onMessage");
+        Type returnType = onMessageResource.getReturnType();
+        Type[] parameterTypes = onMessageResource.getParameterTypes();
+        if (parameterTypes.length == 1) {
+            Object[] args1 = new Object[2];
+            if (parameterTypes[0].getTag() == TypeTags.INTERSECTION_TAG) {
+                args1[0] = getReadonlyMessage(msg);
+            } else {
+                args1[0] = populatedMsgRecord;
+            }
+            args1[1] = true;
+            dispatch(args1, msg.getSubject(), returnType);
+        } else if (parameterTypes.length == 2) {
+            Object[] args2 = new Object[4];
+            if (parameterTypes[0].getTag() == TypeTags.INTERSECTION_TAG) {
+                args2[0] = getReadonlyMessage(msg);
+            } else {
+                args2[0] = populatedMsgRecord;
+            }
+            args2[1] = true;
+            args2[2] = callerObj;
+            args2[3] = true;
+            dispatch(args2, msg.getSubject(), returnType);
+        } else {
+            throw Utils.createNatsError("Invalid remote function signature");
+        }
+    }
+
+    private BMap<BString, Object> getReadonlyMessage(Message msg) {
+        Map<String, Object> valueMap = new HashMap<>();
+        valueMap.put(Constants.MESSAGE_CONTENT, ValueCreator.createArrayValue(msg.getData()));
+        valueMap.put(Constants.MESSAGE_SUBJECT, StringUtils.fromString(msg.getSubject()));
+        return ValueCreator.createReadonlyRecordValue(Utils.getModule(), Constants.STREAM_MESSAGE, valueMap);
+    }
+
+    private void dispatch(Object[] args, String subject, Type returnType) {
+        executeResource(subject, args, returnType);
+    }
+
+    private void executeResource(String subject, Object[] args, Type returnType) {
+        CountDownLatch countDownLatch = new CountDownLatch(1);
```

Comment:
What's the reason for using a CountDownLatch here? Can't we use the async functionality?

Reply:
CountDownLatch is used this way in messaging connectors to ensure that the messages gets received in the same order they are published.